### PR TITLE
Added RecurrentBatchNorm module

### PR DESCRIPTION
**Pull request by rachtsingh**
This PR adds support for recurrent batch normalization to the LSTM and GRU implementations, via the flag `-batch_norm`.

The most important structural change is that the new RecurrentBatchNorm module contains inside it `nn.BatchNormalization` modules, one for each time step. We select the correct one to use in forward/backward passes inside the `:net` function inside `onmt.Sequencer`. This makes the function have side effects, though it should be transparent from the outside.

I'm not 100% sure the implementation for GRU is correct, but the LSTM one does perform better with batch norm on.

### CODE REVIEW

File: onmt/modules/RecurrentBatchNorm.lua [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+  self.momentum = momentum
+  self.affine = affine
+  self.noBias = noBias
+
+  self:setTimeStep(1)
+end
+
+function RecurrentBatchNorm:__timeStepClone()
+  local bn = nn.BatchNormalization(self.dim, self.eps, self.momentum, self.affine)
+  if self._type == 'torch.CudaTensor' then
```
Comment (guillaumekln) [inline onmt/modules/RecurrentBatchNorm.lua]: It is more generic to do:

```lua
bn:type(self._type)
```

And it will also fix CPU execution as we are working with `float`.

File: onmt/modules/Sequencer.lua [hunk 2]
```diff
@@ -26,6 +28,16 @@ end
 function Sequencer:_sharedClone()
   local clone = self.network:clone('weight', 'gradWeight', 'bias', 'gradBias')
 
+  local cloneRNN = onmt.utils.Tensor.findModule(clone, self.rnnClass)
+  if cloneRNN ~= nil then
+    cloneRNN:shareRecurrents(onmt.utils.Tensor.findModule(self.network, self.rnnClass))
+  end
+  function clone:setTimeStep(t)
```
Comment (guillaumekln) [inline onmt/modules/Sequencer.lua]: I think it would be better to just build an array of all RNNs (in `self.rnns`) and define this function statically in the `Sequencer` class.

Also `_sharedClone` is not called at test time so `Sequencer.setTimeStep` will be undefined.

File: onmt/modules/Sequencer.lua [hunk 3]
```diff
@@ -26,6 +28,16 @@ end
 function Sequencer:_sharedClone()
   local clone = self.network:clone('weight', 'gradWeight', 'bias', 'gradBias')
 
+  local cloneRNN = onmt.utils.Tensor.findModule(clone, self.rnnClass)
```
Comment (rachtsingh) [inline onmt/modules/Sequencer.lua]: This is a much better solution! I didn't read the code carefully enough to see this is what was happening. Fixed.

File: onmt/modules/LSTM.lua [hunk 4]
```diff
@@ -104,8 +116,22 @@ function LSTM:_buildLayer(inputSize, hiddenSize)
   local x = inputs[3]
 
   -- Evaluate the input sums at once for efficiency.
-  local i2h = nn.Linear(inputSize, 4 * hiddenSize)(x)
-  local h2h = nn.Linear(hiddenSize, 4 * hiddenSize)(prevH)
+  local i2h = nn.Linear(inputSize, 4 * hiddenSize, false)(x)
```
Comment (rachtsingh) [inline onmt/modules/LSTM.lua]: So first, in the non-batch-normalized case, by having both modules have biases, there's an unnecessary redundancy since they will just be added anyway.

In the batch normalized case, the authors of this [paper](https://arxiv.org/pdf/1603.09025v4.pdf) recommended batch normalizing the output before adding the bias, and having that be separate (see equation 6).

That all being said, I'll measure the differences and report back.

File: onmt/modules/RecurrentBatchNorm.lua (hunk 1 above)
Comment (rachtsingh) [inline onmt/modules/RecurrentBatchNorm.lua]: Sounds great! fixed.

File: onmt/modules/Sequencer.lua (hunk 2 above)
Comment (rachtsingh) [inline onmt/modules/Sequencer.lua]: Ok, I thought that was a little hacky but if its acceptable it's definitely a better solution. 

I overlooked the second fact - that might be what's leading to your NaN perplexity on validation. I will fix this in a second and then push to this PR.

**Comment by rachtsingh** [comment]
Here's a log from running the baseline: https://gist.github.com/rachtsingh/ff6c22a6dfadb41808e9586f8d905163

And the equivalent without batch normalization:
https://gist.github.com/rachtsingh/bdd5712047171f4944c75af01a80af2c

So the validation perplexity doesn't fall faster, but training does. In an earlier experiment it did lead to a 2x smaller validation perplexity after the first epoch, so it might be situation dependent. 

In any case the wall clock speed is much worse, and I'm not entirely sure why - there's a 5% increase in model size, but a 2x decrease in speed?

File: onmt/modules/LSTM.lua (hunk 4 above)
Comment (guillaumekln) [inline onmt/modules/LSTM.lua]: cc @jsenellart-systran

Comment (jsenellart-systran) [inline onmt/modules/LSTM.lua]: for reference, we know that there is a redundancy here - but we added it for being equivalent with cudnn.LSTM. When batch-normalization will be used - the module will not be anymore compatible with cudnn.LSTM anyway, so in that case you can make the change, but let us manage to keep the non-batch-normalization mode as is.

File: onmt/modules/RecurrentBatchNorm.lua [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+  self:setTimeStep(1)
+end
+
+function RecurrentBatchNorm:__timeStepClone()
+  local bn = nn.BatchNormalization(self.dim, self.eps, self.momentum, self.affine)
+  bn:type(self._type)
+  if self.affine then
+    -- initialize weights using the recommended value
+    bn.weight:fill(0.1)
```
Comment (guillaumekln) [inline onmt/modules/RecurrentBatchNorm.lua]: From the paper:

> Note that we separate only the statistics over time and **not the γ and β parameters**.

You may need to call `bn:share(self.modules[1], 'weight', 'gradWeight', 'bias', 'gradBias')` for `t > 1`.

File: onmt/modules/LSTM.lua (hunk 4 above)
Comment (guillaumekln) [inline onmt/modules/LSTM.lua]: So we prefer `not self.batchNorm` instead of `false`.

File: onmt/utils/MemoryOptimizer.lua [hunk 6]
```diff
@@ -14,7 +14,9 @@ local protectInput = {
   'nn.Linear',
   'nn.JoinTable',
   'nn.CMulTable',
-  'nn.MM'
+  'nn.MM',
+  'nn.BatchNormalization',
+  'onmt.RecurrentBatchNorm',
```
Comment (guillaumekln) [inline onmt/utils/MemoryOptimizer.lua]: Just adding `nn.BatchNormalization` is enough as they have both the same input.

**Comment by guillaumekln** [comment]
Did you make more experiments with it? While the implementation seems correct (with parameters sharing), I can't get it to work on a baseline corpus.

Validation perplexity is high, going up and down and predictions are random.
